File: go/pkg/db/exposition.go
package db

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/MaxBlaushild/poltergeist/pkg/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

type expositionHandle struct {
	db *gorm.DB
}

func (h *expositionHandle) preloadBase(ctx context.Context) *gorm.DB {
	return h.db.WithContext(ctx).
		Preload("Zone").
		Preload("PointOfInterest").
		Preload("ItemRewards").
		Preload("ItemRewards.InventoryItem").
		Preload("SpellRewards").
		Preload("SpellRewards.Spell")
}

func (h *expositionHandle) visibleQuery(ctx context.Context) *gorm.DB {
	return h.preloadBase(ctx)
}

func normalizeExpositionRewards(exposition *models.Exposition) {
	if exposition == nil {
		return
	}
	exposition.RequiredStoryFlags = normalizeJSONStringArray(exposition.RequiredStoryFlags)
	if exposition.Dialogue == nil {
		exposition.Dialogue = models.DialogueSequence{}
	}
	if strings.TrimSpace(string(exposition.RewardMode)) == "" {
		if exposition.RewardExperience > 0 ||
			exposition.RewardGold > 0 ||
			len(exposition.ItemRewards) > 0 ||
			len(exposition.SpellRewards) > 0 {
			exposition.RewardMode = models.RewardModeExplicit
		} else {
			exposition.RewardMode = models.RewardModeRandom
		}
	}
	exposition.RewardMode = models.NormalizeRewardMode(string(exposition.RewardMode))
	exposition.RandomRewardSize = models.NormalizeRandomRewardSize(string(exposition.RandomRewardSize))
	if exposition.RewardExperience < 0 {
		exposition.RewardExperience = 0
	}
	if exposition.RewardGold < 0 {
		exposition.RewardGold = 0
	}
}

func (h *expositionHandle) Create(ctx context.Context, exposition *models.Exposition) error {
	if exposition == nil {
		return nil
	}
	normalizeExpositionRewards(exposition)
	return h.db.WithContext(ctx).Create(exposition).Error
}

func (h *expositionHandle) FindByID(ctx context.Context, id uuid.UUID) (*models.Exposition, error) {
	var exposition models.Exposition
	if err := h.preloadBase(ctx).First(&exposition, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return &exposition, nil
}

func (h *expositionHandle) FindAll(ctx context.Context) ([]models.Exposition, error) {
	var expositions []models.Exposition
	if err := h.visibleQuery(ctx).Find(&expositions).Error; err != nil {
		return nil, err
	}
	return expositions, nil
}

func (h *expositionHandle) FindByZoneID(ctx context.Context, zoneID uuid.UUID) ([]models.Exposition, error) {
	var expositions []models.Exposition
	if err := h.visibleQuery(ctx).
		Where("zone_id = ?", zoneID).
		Find(&expositions).Error; err != nil {
		return nil, err
	}
	return expositions, nil
}

func (h *expositionHandle) FindByZoneIDExcludingQuestNodes(ctx context.Context, zoneID uuid.UUID) ([]models.Exposition, error) {
	var expositions []models.Exposition
	if err := h.visibleQuery(ctx).
		Where("zone_id = ?", zoneID).
		Where("NOT EXISTS (SELECT 1 FROM quest_nodes qn WHERE qn.exposition_id = expositions.id)").
		Find(&expositions).Error; err != nil {
		return nil, err
	}
	return expositions, nil
}

func (h *expositionHandle) IsLinkedToQuestNode(ctx context.Context, id uuid.UUID) (bool, error) {
	var count int64
	if err := h.db.WithContext(ctx).
		Table("quest_nodes").
		Where("exposition_id = ?", id).
		Count(&count).Error; err != nil {
		return false, err
	}
	return count > 0, nil
}

func (h *expositionHandle) Update(ctx context.Context, id uuid.UUID, updates *models.Exposition) error {
	if updates == nil {
		return nil
	}
	normalizeExpositionRewards(updates)
	payload := map[string]interface{}{
		"zone_id":               updates.ZoneID,
		"point_of_interest_id":  updates.PointOfInterestID,
		"latitude":              updates.Latitude,
		"longitude":             updates.Longitude,
		"geometry":              updates.Geometry,
		"title":                 updates.Title,
		"description":           updates.Description,
		"dialogue":              updates.Dialogue,
		"required_story_flags":  updates.RequiredStoryFlags,
		"image_url":             updates.ImageURL,
		"thumbnail_url":         updates.ThumbnailURL,
		"reward_mode":           updates.RewardMode,
		"random_reward_size":    updates.RandomRewardSize,
		"reward_experience":     updates.RewardExperience,
		"reward_gold":           updates.RewardGold,
		"material_rewards_json": updates.MaterialRewards,
		"updated_at":            updates.UpdatedAt,
	}
	return h.db.WithContext(ctx).Model(&models.Exposition{}).Where("id = ?", id).Updates(payload).Error
}

func (h *expositionHandle) ReplaceItemRewards(ctx context.Context, expositionID uuid.UUID, rewards []models.ExpositionItemReward) error {
	return h.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		if err := tx.Where("exposition_id = ?", expositionID).Delete(&models.ExpositionItemReward{}).Error; err != nil {
			return err
		}
		now := time.Now()
		for _, reward := range rewards {
			reward.ID = uuid.New()
			reward.ExpositionID = expositionID
			reward.CreatedAt = now
			reward.UpdatedAt = now
			if err := tx.Create(&reward).Error; err != nil {
				return err
			}
		}
		return nil
	})
}

func (h *expositionHandle) ReplaceSpellRewards(ctx context.Context, expositionID uuid.UUID, rewards []models.ExpositionSpellReward) error {
	return h.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		if err := tx.Where("exposition_id = ?", expositionID).Delete(&models.ExpositionSpellReward{}).Error; err != nil {
			return err
		}
		now := time.Now()
		for _, reward := range rewards {
			reward.ID = uuid.New()
			reward.ExpositionID = expositionID
			reward.CreatedAt = now
			reward.UpdatedAt = now
			if err := tx.Create(&reward).Error; err != nil {
				return err
			}
		}
		return nil
	})
}

func (h *expositionHandle) UpsertCompletion(ctx context.Context, userID uuid.UUID, expositionID uuid.UUID) error {
	now := time.Now()
	record := models.UserExpositionCompletion{
		ID:           uuid.New(),
		CreatedAt:    now,
		UpdatedAt:    now,
		UserID:       userID,
		ExpositionID: expositionID,
	}
	return h.db.WithContext(ctx).
		Clauses(clause.OnConflict{
			Columns:   []clause.Column{{Name: "user_id"}, {Name: "exposition_id"}},
			DoUpdates: clause.Assignments(map[string]interface{}{"updated_at": now}),
		}).
		Create(&record).Error
}

func (h *expositionHandle) FindCompletionByUserAndExposition(ctx context.Context, userID uuid.UUID, expositionID uuid.UUID) (*models.UserExpositionCompletion, error) {
	var completion models.UserExpositionCompletion
	if err := h.db.WithContext(ctx).
		Where("user_id = ? AND exposition_id = ?", userID, expositionID).
		First(&completion).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return &completion, nil
}

func (h *expositionHandle) FindCompletedExpositionIDsByUser(ctx context.Context, userID uuid.UUID, expositionIDs []uuid.UUID) ([]uuid.UUID, error) {
	if len(expositionIDs) == 0 {
		return nil, nil
	}
	var completedIDs []uuid.UUID
	if err := h.db.WithContext(ctx).
		Model(&models.UserExpositionCompletion{}).
		Where("user_id = ?", userID).
		Where("exposition_id IN ?", expositionIDs).
		Pluck("exposition_id", &completedIDs).Error; err != nil {
		return nil, err
	}
	return completedIDs, nil
}

func (h *expositionHandle) Delete(ctx context.Context, id uuid.UUID) error {
	return h.db.WithContext(ctx).Delete(&models.Exposition{}, "id = ?", id).Error
}

File: dart/unclaimed_streets/lib/models/exposition.dart
import 'character_action.dart';
import 'point_of_interest.dart';
import 'spell.dart';

class Exposition {
  final String id;
  final String zoneId;
  final String? pointOfInterestId;
  final PointOfInterest? pointOfInterest;
  final double latitude;
  final double longitude;
  final String title;
  final String description;
  final List<DialogueMessage> dialogue;
  final String imageUrl;
  final String thumbnailUrl;
  final String rewardMode;
  final String randomRewardSize;
  final int rewardExperience;
  final int rewardGold;
  final List<Map<String, dynamic>> materialRewards;
  final List<Map<String, dynamic>> itemRewards;
  final List<Spell> spellRewards;

  const Exposition({
    required this.id,
    required this.zoneId,
    this.pointOfInterestId,
    this.pointOfInterest,
    required this.latitude,
    required this.longitude,
    required this.title,
    required this.description,
    this.dialogue = const [],
    this.imageUrl = '',
    this.thumbnailUrl = '',
    this.rewardMode = 'random',
    this.randomRewardSize = 'small',
    this.rewardExperience = 0,
    this.rewardGold = 0,
    this.materialRewards = const [],
    this.itemRewards = const [],
    this.spellRewards = const [],
  });

  factory Exposition.fromJson(Map<String, dynamic> json) {
    PointOfInterest? pointOfInterest;
    final rawPoi = json['pointOfInterest'];
    if (rawPoi is Map<String, dynamic>) {
      pointOfInterest = PointOfInterest.fromJson(rawPoi);
    } else if (rawPoi is Map) {
      pointOfInterest = PointOfInterest.fromJson(
        Map<String, dynamic>.from(rawPoi),
      );
    }

    final dialogue = <DialogueMessage>[];
    final rawDialogue = json['dialogue'];
    if (rawDialogue is List) {
      for (final entry in rawDialogue) {
        if (entry is Map<String, dynamic>) {
          dialogue.add(DialogueMessage.fromJson(entry));
        } else if (entry is Map) {
          dialogue.add(
            DialogueMessage.fromJson(Map<String, dynamic>.from(entry)),
          );
        }
      }
    }

    final spellRewards = <Spell>[];
    final rawSpellRewards = json['spellRewards'];
    if (rawSpellRewards is List) {
      for (final reward in rawSpellRewards) {
        if (reward is Map<String, dynamic>) {
          final rawSpell = reward['spell'];
          if (rawSpell is Map<String, dynamic>) {
            spellRewards.add(Spell.fromJson(rawSpell));
          } else if (rawSpell is Map) {
            spellRewards.add(
              Spell.fromJson(Map<String, dynamic>.from(rawSpell)),
            );
          }
        } else if (reward is Map) {
          final rawSpell = reward['spell'];
          if (rawSpell is Map<String, dynamic>) {
            spellRewards.add(Spell.fromJson(rawSpell));
          } else if (rawSpell is Map) {
            spellRewards.add(
              Spell.fromJson(Map<String, dynamic>.from(rawSpell)),
            );
          }
        }
      }
    }

    return Exposition(
      id: json['id']?.toString() ?? '',
      zoneId: json['zoneId']?.toString() ?? '',
      pointOfInterestId: json['pointOfInterestId']?.toString(),
      pointOfInterest: pointOfInterest,
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
      title: json['title']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      dialogue: dialogue,
      imageUrl: json['imageUrl']?.toString() ?? '',
      thumbnailUrl: json['thumbnailUrl']?.toString() ?? '',
      rewardMode: json['rewardMode']?.toString() ?? 'random',
      randomRewardSize: json['randomRewardSize']?.toString() ?? 'small',
      rewardExperience: (json['rewardExperience'] as num?)?.toInt() ?? 0,
      rewardGold: (json['rewardGold'] as num?)?.toInt() ?? 0,
      materialRewards:
          (json['materialRewards'] as List<dynamic>?)
              ?.whereType<Map>()
              .map((entry) => Map<String, dynamic>.from(entry))
              .toList() ??
          const [],
      itemRewards:
          (json['itemRewards'] as List<dynamic>?)
              ?.whereType<Map>()
              .map((entry) => Map<String, dynamic>.from(entry))
              .toList() ??
          const [],
      spellRewards: spellRewards,
    );
  }
}

class ExpositionPerformResult {
  final String expositionId;
  final bool successful;
  final String title;
  final int rewardExperience;
  final int rewardGold;
  final List<Map<String, dynamic>> baseResourcesAwarded;
  final List<Map<String, dynamic>> itemsAwarded;
  final List<Spell> spellsAwarded;
  final bool awardedRewards;

  const ExpositionPerformResult({
    required this.expositionId,
    required this.successful,
    required this.title,
    required this.rewardExperience,
    required this.rewardGold,
    this.baseResourcesAwarded = const [],
    this.itemsAwarded = const [],
    this.spellsAwarded = const [],
    this.awardedRewards = false,
  });

  factory ExpositionPerformResult.fromJson(Map<String, dynamic> json) {
    final spellsAwarded =
        (json['spellsAwarded'] as List<dynamic>?)
            ?.whereType<Map>()
            .map((spell) => Spell.fromJson(Map<String, dynamic>.from(spell)))
            .toList() ??
        const <Spell>[];

    return ExpositionPerformResult(
      expositionId: json['expositionId']?.toString() ?? '',
      successful: json['successful'] == true,
      title: json['title']?.toString() ?? '',
      rewardExperience: (json['rewardExperience'] as num?)?.toInt() ?? 0,
      rewardGold: (json['rewardGold'] as num?)?.toInt() ?? 0,
      baseResourcesAwarded:
          (json['baseResourcesAwarded'] as List<dynamic>?)
              ?.whereType<Map>()
              .map((entry) => Map<String, dynamic>.from(entry))
              .toList() ??
          const [],
      itemsAwarded:
          (json['itemsAwarded'] as List<dynamic>?)
              ?.whereType<Map>()
              .map((entry) => Map<String, dynamic>.from(entry))
              .toList() ??
          const [],
      spellsAwarded: spellsAwarded,
      awardedRewards: json['awardedRewards'] == true,
    );
  }
}
